### DISCUSSION TITLE
Valetudo 2024.10.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#2114] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2024.10.0</h2></p>
</div>

Obstacle Images - a surprisingly fun feature

## Obstacle Images

On robots that support the feature, assuming the right zoom level, you will now be able to click on the obstacle in the map view and see what the robot saw. I didn't expect it to be, but it is _very neat_ to view the world through the perspective of the robot.

![image](https://github.com/user-attachments/assets/3fc50b35-5d2e-4b6a-bff9-4fbc0486a75e)

I was quite hesitant to implement such a feature for a long time, because there is a lot that one can do wrong when reading arbitrary large files based on user input from the filesystem and delivering them via HTTP.
But, given that long time I had to think about it, I believe that the implementation is now solid.

![image](https://github.com/user-attachments/assets/7c17f89a-5e89-4d77-a4cc-58a9642e1c87)

I hope that you will have as much fun with this feature as I had.
And, if your robot sees something neat and shareable, feel free to post the obstacle image in the comments below.

**Side-note**:<br/>
This should go without saying but please don't use this feature for abuse.
Talk with everyone living in (or even just visiting) your household about this **before** you enable it.

Even if you live alone, you can be certain that the robot will take a picture at the most unfortunate and inopportune moment imaginable.
Keep that in mind and maybe just keep the feature turned off if unsure.

The obstacle avoidance works regardless of it, so you're not losing out on any functionality if you do decide to not enable it.

## Improvements to the Virtual Restrictions Editor

A simple change with great impact to the UX:<br/>
The virtual restrictions editor now also renders the path

![image](https://github.com/user-attachments/assets/f941b89f-08d5-46d1-a545-0f8f368348ba)

This makes it a _lot_ easier to precisely place virtual restrictions based on the real struggles the robot faced.

## UI UX Fixes

Cutting lines for segments now have a shadow to not become invisible in front of the white background of the light theme.

![image](https://github.com/user-attachments/assets/0e682533-a350-4557-b644-f76f140bc79f)

Furthermore, zooming the map in by 1 step and zooming back out by 1 step will now make you end up with the same zoom level you started with - as one would expect it to.

Thanks to @vivia for both of these!

## Improvements to the System Information View

The System Information page now displays the actual CPU load:

![image](https://github.com/user-attachments/assets/bcf24a5a-d6db-401d-95bc-084b45fc6a6a)

Previously, it used the system load, which got very confusing on robots with kernels where that feature is broken for some reason.
Now, these bars make sense again instead of just being stuck at 100% all the time.

## Misc

1. Users of the Dreame L10s Pro Ultra Heat and X40 can now trigger the Mop-Dock cleaning procedure using a new quirk.

    ![image](https://github.com/user-attachments/assets/d47d06b5-8477-4285-9508-630f580cb02c)

2. Non-feature-wise, this release comes with a bump to basically all dependencies, meaning that we're good on that front as well again.

## The usual

If you want to see Valetudo on more robots and/or like this release, you might want to consider donating:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Features

- **vendor.dreame**: D10s Pro/Plus also support Obstacle Images [`63279c4`](http://github.com/Hypfer/Valetudo/commit/63279c44a0362e8b75e9729c63c655a6d7863089)
- **vendor.dreame**: Add quirk for triggering the mop dock cleaning procedure [`76280bd`](http://github.com/Hypfer/Valetudo/commit/76280bdfc6c75bdfb6a6be72c53e631a35b60f8c)
- **vendor.dreame**: ObstacleImagesCapability [`67ee71c`](http://github.com/Hypfer/Valetudo/commit/67ee71ce58528e78eb0664f4980ede54f5625b28)
- **ui**: ObstacleImagesCapability [`30b5fc8`](http://github.com/Hypfer/Valetudo/commit/30b5fc8a3d248e16c7868d3343688632a4ca3435)
- **core**: ObstacleImagesCapability [`f9fa9d0`](http://github.com/Hypfer/Valetudo/commit/f9fa9d0b6027ae188fe219faa5b24e9ae840bfa1)
- Display the actual CPU usage instead of the system load [`d0f754b`](http://github.com/Hypfer/Valetudo/commit/d0f754bc1029b10859648980349fb5bfacb21505)
- **ui**: Draw half-transparent path in virtual restrictions edit map to make placing restrictions correctly easier [`ca077b6`](http://github.com/Hypfer/Valetudo/commit/ca077b66c345599293b721afadf40444945838b9)

### Fixes

- **ui**: Fix zoom level after zooming in and back out again not ending up where it started [`2e5a9c2`](http://github.com/Hypfer/Valetudo/commit/2e5a9c20bcdb19e43fecb5a62c99027c8dcaf734)
- **vendor.dreame**: Fix MopDockCleanManualTriggerCapability for everything that is not a pure mop [`3513cbc`](http://github.com/Hypfer/Valetudo/commit/3513cbc698cdc27fee86c9dda12f0f98e5eba957)
- **webserver**: Valetudo router rate limits should be global [`7a6eb1e`](http://github.com/Hypfer/Valetudo/commit/7a6eb1ee0b85bbafd7776d513b505ba2f01f1dd1)
- **vendor.dreame**: Fix Mop Dock Water Heater quirk [`20917a7`](http://github.com/Hypfer/Valetudo/commit/20917a705691586f369e511d9eeb3e560335ba58)
- A config reset should not reset the robot config [`28d674f`](http://github.com/Hypfer/Valetudo/commit/28d674f4e5d3dcda97807c855c018c35e1f7f9bb)
- **ui**: Fix cutting line being invisible when using the light theme [`a2a1d40`](http://github.com/Hypfer/Valetudo/commit/a2a1d4061d605bd60c148a6464a1945ff59595bb)

### Refactoring

- **vendor.dreame**: Add more AI classifier IDs and pull the constant into DreameUtils [`0928cf0`](http://github.com/Hypfer/Valetudo/commit/0928cf01aff8b374dd9759a5252b6b63b9dc33a5)
- **webserver**: Use rate limiters instead of semaphore in ObstacleImagesCapabilityRouter [`90e5ab9`](http://github.com/Hypfer/Valetudo/commit/90e5ab935fc8a9d73365f0babcaab0b93054bf2e)

### Chores

- Bump dependencies [`5559f88`](http://github.com/Hypfer/Valetudo/commit/5559f88ab2a3367603f904a164f5a388bc7ccd98)
- Upgrade react-router v5 to v6 [`fb600a8`](http://github.com/Hypfer/Valetudo/commit/fb600a8ef5ab65e208e20ff1ddf8a77a85388bc0)
- Bump dependencies [`79deca1`](http://github.com/Hypfer/Valetudo/commit/79deca12ea907228d40298ed6fff1a133ae780e3)
- Fix the wrong link in the right place [`2152e20`](http://github.com/Hypfer/Valetudo/commit/2152e20b6eb34eed0bad02cd3fba6f34a1495a86)
- Bump some transitive dependencies [`dc3b96e`](http://github.com/Hypfer/Valetudo/commit/dc3b96e87d0f9a3c5fdcacf299ab33d42bfad691)
- **ui**: Resolve issues detected by sonarcloud [`b56ed1c`](http://github.com/Hypfer/Valetudo/commit/b56ed1ca6e29cde2ad2326c5bbc8c81a9b6146e6)
- Update PR template [`ddcf3e1`](http://github.com/Hypfer/Valetudo/commit/ddcf3e15909dce5e49815422566b8a8fbe6f48da)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2024.10.0'>Valetudo 2024.10.0</a>.</em>

**Comment by rouke-broersma** [comment]
> The virtual restrictions editor now also renders the path

Amazing! I have been needing this feature for so long!

**Comment by csidirop** [comment]
I assume, that the Obstacle Image option should be right after the Obstacle Avoidance right? And if its not there the robot does not support it?

![Screenshot_20241024-095004_Firefox_small](https://github.com/user-attachments/assets/52438c75-9d6c-4a05-af13-111b8b5bf571)


I have a Dreame L10 Pro

**Reply by Hypfer** [reply]
Yeah it would be somewhere around there.
It's a feature of the AI Camera Obstacle Avoidance. Your L10 only features Line Laser Obstacle Avoidance

Even if the firmware of your bot would take and provide such pictures, they'd only look like this:

![image](https://github.com/user-attachments/assets/a4a35080-6491-4500-9d24-735be1f328d5)


Where newer robots do object recognition in images, line laser just projects an IR laser cross line in front of the robot (below where the lidar can see) and checks if that is straight or somehow disturbed

**Reply by csidirop** [reply]
Okay thank you

**Comment by justin-sleep** [comment]
Loving the feature so far!

![e47fc669-eb01-5e73-b316-626fe4e6c5c6](https://github.com/user-attachments/assets/dbe22969-f7c8-4eb0-b02f-6d2594b69ddd)

![08d38836-2966-5bd0-94ab-88a86cb7ea2b](https://github.com/user-attachments/assets/10fb399d-eed7-4655-866f-f864754ce25d)
